gitlab_group: fix issue 4577

fzarifian opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
Fix https://github.com/ansible-collections/community.general/issues/4577

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

felixfontein commented on plugins/modules/source_control/gitlab/gitlab_group.py:
```diff
@@ -62,6 +62,7 @@
     description:
       - Determine if developers can create projects in the group.
     choices: ["developer", "maintainer", "noone"]
+    default: "maintainer"
```
Defaults are not acceptable. They cause problems with idempotence (tasks suddenly changing properties of objects that were not specified).

fzarifian replied:
not exactly, an idempotency problem exists if you can create it... and for now, it's impossible.

but i understand you point of view : on already created (by hand) this one can change the attribute (because i dont take time to find the default value gitlab choose, feel free to change this default value)

And for the last question : those two attributes are mandatory for creation, but not for update

fzarifian replied:
and for the last one : feel free to change the documentation to add those mandatories options for creation and reject this PR it does work too.

fzarifian replied:
the last commit resolve the discussion

felixfontein reviewed (changes requested):
Thanks for your contribution. Unfortunately your fix does cause new bugs and cannot be merged in this form.

Lunik commented:
If we can't set à default value but the API expect those values, we should mark them as `required`

Lunik commented:
Can also fix #4577

felixfontein commented:
You can also only require it on creation (and fail with a nicer error message), or dynamically pass a default on creation if no value is passed to the module.

fzarifian commented:
@felixfontein the value is not mandatory, but passing the argument as null fails

felixfontein commented:
Then forcing a default when one isn't needed is really a wrong fix. The correct fix would be to prevent `null` being passed when no value is specified.

fzarifian commented:
```python
    def create_or_update_group(self, name, parent, options):
        changed = False

        payload = {
            'name': name,
            'path': options['path'],
            'visibility': options['visibility'],
            'project_creation_level': options['project_creation_level'],
            'auto_devops_enabled': options['auto_devops_enabled'],
            'subgroup_creation_level': options['subgroup_creation_level'],
        }
        if options.get('project_creation_level'):
            payload['project_creation_level'] = options['project_creation_level']
        if options.get('subgroup_creation_level'):
            payload['subgroup_creation_level'] = options['subgroup_creation_level']
        if options.get('description'):
            payload['description'] = options['description']
        if options.get('require_two_factor_authentication'):
            payload['require_two_factor_authentication'] = options['require_two_factor_authentication']

        # Because we have already call userExists in main()
        if self.group_object is None:
            parent_id = self.get_group_id(parent)
            payload['parent_id'] = parent_id
            group = self.create_group(payload)

            # add avatar to group
            if options['avatar_path']:
                try:
                    group.avatar = open(options['avatar_path'], 'rb')
                except IOError as e:
                    self._module.fail_json(msg='Cannot open {0}: {1}'.format(options['avatar_path'], e))
            changed = True
        else:
            changed, group = self.update_group(self.group_object, payload)

        self.group_object = group
        if changed:
            if self._module.check_mode:
                self._module.exit_json(changed=True, msg="Successfully created or updated the group %s" % name)

            try:
                group.save()
            except Exception as e:
                self._module.fail_json(msg="Failed to update group: %s " % e)
            return True
        else:
            return False
```

but continue to fail... from my point of view, i have no other or better solution to propose, feel free to close it.

Another way to do is to update the documentation : 

```
- name: "Create GitLab SubGroup"
  community.general.gitlab_group:
    api_url: https://gitlab.example.com/
    validate_certs: True
    api_username: dj-wasabi
    api_password: "MySecretPassword"
    name: my_first_group
    path: my_first_group
    state: present
    parent: "super_parent/parent"
```

fzarifian commented:
as people does not like this PR, feel free to only update documentation adding those two arguments.

fzarifian commented:
Some typos in my payload, remaining `'subgroup_creation_level': options['subgroup_creation_level'],` after putting it in optional : works great :)

This one resolve :
- the broken creation 
- the idempotency for those optionnal values

After that, looots of questions about the avatar implemention, but that's not the PR subject...

fzarifian commented:
the implementation of the check_mode is also interresting, but not the PR subject too

felixfontein commented on plugins/modules/source_control/gitlab/gitlab_group.py:
```diff
@@ -197,23 +197,27 @@ def get_group_id(self, group):
+        # Same payload for update or create, only parent_id is needed on creation
+        payload = {
+            'name': name,
+            'path': options['path'],
```
So far `path` was only passed on creation, I would handle it the same as `parent_id`.

felixfontein reviewed (commented):
Besides that, looks good. (Still needs a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment) though.)

fzarifian commented:
In reality, only two parameters are mandatory in gitlab implementation.
- name
- path (slugified name, which should be a method defined in module_utils, and not the code used because it does not handle the lowercase needed in slugify, that’s related to the think you can see in the main method ; I agree that the payload of the argspec should be used in the whole module rather than this implementation with string mapping 3 times in this module); note 

- all others are not mandatory.

https://docs.gitlab.com/ee/api/groups.html#new-group

I think this is more than this PR to review the whole code and I don’t have time to do it for now, feel free to assign me the change request to refactor it

To close this one, I don’t undestand the changelog fragment very well, I need to read the documentation you put. Will come back with new submission when it’s ok for me.

— 
Fabien ZARIFIAN

> Le 20 juin 2022 à 07:51, Felix Fontein ***@***.***> a écrit :
> 
> ﻿
> @felixfontein commented on this pull request.
> 
> Besides that, looks good. (Still needs a changelog fragment though.)
> 
> In plugins/modules/source_control/gitlab/gitlab_group.py:
> 
> > @@ -197,23 +197,27 @@ def get_group_id(self, group):
>      def create_or_update_group(self, name, parent, options):
>          changed = False
>  
> +        # Same payload for update or create, only parent_id is needed on creation
> +        payload = {
> +            'name': name,
> +            'path': options['path'],
> So far path was only passed on creation, I would handle it the same as parent_id.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

eifelmicha commented:
@fzarifian Just stumbled across this. Before switching over to curl requests. What's the status of this fix?

felixfontein commented:
Please note that in #5461 the collection repository was restructured to remove the directory tree in plugins/modules/, and the corresponding tree in tests/unit/plugins/modules/. Your PR modifies files in this directory structure, and unfortunately now has some conflicts, potentially because of this. Please **rebase** with the current `main` branch to remove the conflicts.

felixfontein commented:
needs_info
